File: WEEK1/woohyun/파괴되지 않은 건물.js
// 나의 풀이 (정확성 10/10 통과, 효율성 0/7 통과)
function solution(board, skill) {
    var answer = 0;
    
    for (eachSkill of skill) {
        for (var row = eachSkill[1]; row<=eachSkill[3]; row++) {
            for(var col = eachSkill[2]; col <= eachSkill[4]; col++) {
                board[row][col] += eachSkill[5] * (eachSkill[0] == 1 ? -1 : 1);
            }
        }
    }
    
    for (row of board) {
        for(element of row) {
            if(element > 0) answer++;
        }
    }
    
    return answer;
}

// 효율성 풀이 모범 답안 (누적합 개념 사용)

var damages; // 건물에 가해진 데미지의 총합만 구하는 행렬 (누적합을 위해 board 배열보다 가로, 세로의 길이가 1만큼 길다)

function solution(board, skill) {
    var answer = 0;
    
    // damages 배열 전부 0으로 초기화
    damages = Array.from(
        Array(board.length + 1),
        (_) => Array(board[0].length+1).fill(0),
    );
    
    // damages 배열에 누적합을 위한 모서리 초기화
    skill.forEach((eachSkill) => {
        const [type, r1, c1, r2, c2, degree] = eachSkill;
        const damage = degree * (type === 1 ? -1 : 1);
        
        damages[r1][c1] += damage;
        damages[r1][c2+1] += -damage;
        damages[r2+1][c1] += -damage;
        damages[r2+1][c2+1] += damage;
    });
    
    // damages 가로합 (오른쪽으로)
    for(row of damages) {
        for (var col=1; col < row.length; col++) {
            row[col] += row[col-1];
        }
    }
    
    // damages 세로합 (아래로)
    for(var row=1; row < damages.length; row++) {
        for (var col=0; col < damages[row].length; col++) {
            damages[row][col] += damages[row-1][col];
        }
    }
    
    // board의 크기만큼 damages와 비교하여 정답 도출
    for(var row=0; row < board.length; row++) {
        for (var col=0; col < board[row].length; col++) {
            const density = board[row][col];
            const damage = damages[row][col];
            
            if(density + damage > 0) answer++;
        }
    }
    
    return answer;
}

File: WEEK1/yuncheol/파괴되지 않은 건물.java
class Solution {
    public int solution(int[][] board, int[][] skill) {
        int answer = 0;
        
        int R = board.length;
        int C = board[0].length;
        int[][] prefixSum = new int[R+1][C+1];
        
        for (int[] element: skill) {
            int type = element[0];
            int r1 = element[1];
            int c1 = element[2];
            int r2 = element[3];
            int c2 = element[4];
            int degree = element[5];
            
            action(type, r1, c1, r2, c2, degree, prefixSum);
        }
        
        // prefix sum (left to right)
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                prefixSum[r][c+1] += prefixSum[r][c];
            }
        }
        
        // prefix sum (top to bottom)
        for (int c = 0; c < C; c++) {
            for (int r = 0; r < R; r++) {
                prefixSum[r+1][c] += prefixSum[r][c];
            }
        }
        
        // board + prefix sum
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                board[r][c] += prefixSum[r][c];
                
                if (board[r][c] > 0) {
                    answer += 1;
                }
            }
        }
        
        return answer;
    }
    
    private void action(int type, int r1, int c1, int r2, int c2, int degree, int[][] prefixSum) {
        if (type == 1) {
            type = -1;
        } else if (type == 2) {
            type = 1;
        }
        
        prefixSum[r1][c1] += degree * type;
        prefixSum[r1][c2+1] += (-1) * degree * type;
        prefixSum[r2+1][c1] += (-1) * degree * type;
        prefixSum[r2+1][c2+1] += degree * type;
    }
}